src/Engine/Tables/ItemTable.cpp

```diff
@@ -365,182 +365,171 @@ void ItemTable::LoadPotionNotes(const Blob &potionNotes) {
+            int pickedWeight = grng->random(weightSum) + 1;
+            auto foundWeight = std::lower_bound(cumulativeWeights.begin(), cumulativeWeights.end(), pickedWeight);
+
+            assert(foundWeight != cumulativeWeights.end());
+
+            outItem->uItemID = possibleItems[std::distance(cumulativeWeights.begin(), foundWeight)];
         } else {
             outItem->uItemID = ITEM_CRUDE_LONGSWORD;
         }
     } else {
-        // artifact
-        if (treasure_level == ITEM_TREASURE_LEVEL_6) {
-            int artifactsFound = 0;  // [sp+CA0h] [bp+Ch]@55
+        // Trying to generate artifact
+        if (treasureLevel == ITEM_TREASURE_LEVEL_6) {
+            int artifactsFound = 0;
+            ItemId artifactRandomId = grng->randomSample(allSpawnableArtifacts());
             for (ItemId i : allSpawnableArtifacts())
                 artifactsFound += pParty->pIsArtifactFound[i];
-            artifactRandomId = grng->randomSample(allSpawnableArtifacts());
-            if ((grng->random(100) < 5) && !pParty->pIsArtifactFound[artifactRandomId] &&
-                (engine->config->gameplay.ArtifactLimit.value() == 0 || artifactsFound < engine->config->gameplay.ArtifactLimit.value())) {
-                pParty->pIsArtifactFound[artifactRandomId] = 1;
+            bool artifactLimitReached = (engine->config->gameplay.ArtifactLimit.value() != 0 && artifactsFound >= engine->config->gameplay.ArtifactLimit.value());
+            if ((grng->random(100) < 5) && !pParty->pIsArtifactFound[artifactRandomId] && !artifactLimitReached) {
+                pParty->pIsArtifactFound[artifactRandomId] = true;
                 outItem->uAttributes = 0;
                 outItem->uItemID = artifactRandomId;
                 SetSpecialBonus(outItem);
                 return;
             }
         }
 
-        v57 = 0;
-        v18 = grng->random(this->chanceByTreasureLevelSums[treasure_level]) + 1;
-        while (v57 < v18) {
-            // TODO(captainurist): what's going on here? Get rid of casts.
-            outItem->uItemID = ItemId(std::to_underlying(outItem->uItemID) + 1);
-            v57 += pItems[outItem->uItemID].uChanceByTreasureLvl[treasure_level];
+        // Otherwise try to spawn any random item
+        int randomWeight = grng->random(this->chanceByTreasureLevelSums[treasureLevel]) + 1;
+        for (ItemId itemId : allSpawnableItems()) {
+            weightSum += pItems[itemId].uChanceByTreasureLvl[treasureLevel];
+            if (weightSum >= randomWeight) {
+                outItem->uItemID = itemId;
+                break;
+            }
         }
     }
     if (outItem->isPotion() && outItem->uItemID != ITEM_POTION_BOTTLE) {  // if it potion set potion spec
-        outItem->potionPower = 0;
-        for (int i = 0; i < 2; ++i) {
-            outItem->potionPower += grng->random(4) + 1;
-        }
-        outItem->potionPower = outItem->potionPower * std::to_underlying(treasure_level);
+        // Generate "2d4 + 2" value
```

Comment:
That's actually just 2d4. You can replace the call with `grng->randomDice(2, 4)`.

Reply:
Oops, yeah. Didn't thought to look for this function.